docs: add a testing cheat code file (#302)
dojoengine/book: docs/pages/framework/testing-cheat-codes.md

# Cairo Testing Cheat Codes: A Comprehensive Guide

The Cairo Testing Cheat Codes allow you to set and manipulate various execution context variables, such as block number, caller address, contract address, to test your contracts in different scenarios. In this guide, we will explore each cheat code in detail, providing explanations and examples to help you understand how to use them effectively in your tests.

1. `set_block_number`

This cheat code helps you set the current block number to the specified value, allowing you to simulate different block heights for testing purposes.

> - Testing a contract's behavior at a specific block height, such as checking if a certain function is only callable after a certain block number.

Reviewer: ```suggestion
This one is helpful to test a contract's behavior at a specific block height, such as checking if a certain function is only callable after a certain block number.
```

> - Simulating a fork or replay attack to test a contract's resilience.

Reviewer: ```suggestion
```

```
use starknet::{testing, get_block_number};

testing::set_block_number(1234567);

> get_block_nunmber(); // returns 1234567

Reviewer: ```suggestion
assert!(get_block_number() == 1234567, 'bad block number');
```

```

2. `set_caller_address`

This cheat code helps you set the caller address to the provided contract address, enabling you to test contract interactions with different callers. This cheat code can be applied when:

- Testing a contract's access control mechanisms, such as only allowing certain addresses to call specific functions.
- Simulating a scenario where a contract is called by a different address.

```

> use core::traits::{TryInto, Into};
> use starknet::{testing, get_caller_address};
> 
> const USER_ONE: felt252 = 'USER1';
> testing::set_caller_address(USER_ONE.try_into().unwrap()); // Sets user to USER_ONE
> 
> get_caller_address(); // returns USER_ONE.try_into().unwrap()
> 

Reviewer: ```suggestion
use starknet::{testing, get_caller_address, contract_address_const};

let user_one = contract_address_const::<'user1'>();
testing::set_caller_address(user_one);
assert(get_caller_address() == user_one, 'bad caller';
```

```

3. `set_contract_address`

> This cheat code helps you set the contract address to the provided value, allowing you to test contract deployment and interactions. You can apply this code when:

Reviewer: ```suggestion
This cheat code helps you set the contract address to the provided value, allowing you to test contract deployment and interactions.
```

> - Deploying a contract to a specific address for testing purposes.
> - Testing a contract's behavior when deployed to a different address.

Reviewer: ```suggestion
It is important to note that any test function is considered a contract, which by default uses the `0` address. Using `set_contract_address` allows you to mock the current address of the testing function, making it useful to call other contract that may use `get_caller_address`.
```

```

> use core::traits::{TryInto, Into};

Reviewer: ```suggestion
```
Shouldn't be necessary.
